samples/onnx/OnnxGenerator.java

```diff
@@ -0,0 +1,131 @@
+    private final Consumer<String> out;
+
+    public OnnxGenerator(String modelPath, Consumer<String> out) {
+        arena = Arena.ofConfined();
+        ret = arena.allocate(C_POINTER);
+        model = call(OgaCreateModel(arena.allocateFrom(modelPath), ret));
+        tokenizer = call(OgaCreateTokenizer(model, ret));
+        tokenizerStream = call(OgaCreateTokenizerStream(tokenizer, ret));
+        generatorParams = call(OgaCreateGeneratorParams(model, ret));
+        call(OgaGeneratorParamsSetSearchNumber(generatorParams, arena.allocateFrom("max_length"), 1024)); // this is necessary to run with GPU model
+        generator = call(OgaCreateGenerator(model, generatorParams, ret));
+        count = arena.allocate(C_LONG);
+        this.out = out;
+    }
+
+    public static void main(String[] args) throws Exception {
+        System.loadLibrary("onnxruntime-genai");
+        Reader inreader = new InputStreamReader(System.in);
+        try (var gen = new OnnxGenerator(args[0], System.out::print)) {
+            BufferedReader in = new BufferedReader(inreader);
+            String str;
+            System.out.print("> ");
+            while ((str = in.readLine()) != null) {
+                gen.prompt(str);
+                System.out.print("> ");
+            }
+            in.close();
+        }
+    }
+
+    private MemorySegment call(MemorySegment status) {
+        try {
+            if (!status.equals(MemorySegment.NULL)) {
+                status = status.reinterpret(C_INT.byteSize());
+                if (status.get(C_INT, 0) != 0) {
+                    String errString = OgaResultGetError(status)
+                            .reinterpret(Long.MAX_VALUE)
+                            .getString(0L);
+                    throw new RuntimeException(errString);
+                }
+            }
+            return ret.get(C_POINTER, 0);
+        } finally {
+            OgaDestroyResult(status);
+        }
+    }
+
+    public int prompt(String userPrompt) {
+        var inputTokens = call(OgaCreateSequences(ret));
+        int ntokens = 0;
+        try {
+            call(OgaTokenizerEncode(tokenizer, arena.allocateFrom(PROMPT_TEMPLATE.formatted(userPrompt)), inputTokens));
+            call(OgaGenerator_AppendTokenSequences(generator, inputTokens));
+            while (!OgaGenerator_IsDone(generator)) {
+                ntokens++;
+                call(OgaGenerator_GenerateNextToken(generator));
+                int nextToken = call(OgaGenerator_GetNextTokens(generator, ret, count)).get(C_INT, 0);
+                String response = call(OgaTokenizerStreamDecode(tokenizerStream, nextToken, ret)).getString(0);
+                out.accept(response);
+            }
+            out.accept("\n");
+            return ntokens;
+        } finally {
+            OgaDestroySequences(inputTokens);
+        }
+    }
+
+    @Override
+    public void close() throws Exception {
+        arena.close();
+        OgaDestroyGenerator(generator);
```

Comment:
I wonder if we should `MemorySegment::reinterpret` to associate cleanup action to the various segments -- and then just call `arena.close()` from here?

Reply:
(this could be achieved by passing a cleanup action to `call` and then calling reinterpret from there with the cleanup action)